### PR TITLE
[Service Bus] Performance Tests

### PR DESCRIPTION
**Pull request by kashifkhan**
Enable Performance Tests for Service Bus

* Deleted T1 Tests
* The tests now use the same queue for sending and receiving. For receives, preload checks to see messages already present and only adds in the delta - similar to what EH does

### CODE REVIEW

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 1]
```diff
@@ -5,124 +5,210 @@
...
-from azure_devtools.perfstress_tests import PerfStressTest, get_random_bytes
+from azure_devtools.perfstress_tests import PerfStressTest, get_random_bytes, BatchPerfTest
 
-from azure.servicebus import ServiceBusClient, ServiceBusReceiveMode, ServiceBusMessage
+from azure.servicebus import ServiceBusClient, ServiceBusReceiveMode, ServiceBusMessage, TransportType
 from azure.servicebus.aio import ServiceBusClient as AsyncServiceBusClient
 from azure.servicebus.aio.management import ServiceBusAdministrationClient
 
-MAX_QUEUE_SIZE = 40960
+MAX_QUEUE_SIZE=4096
```
Comment (l0lawrence) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:14]: Question: Why is this smaller is it b/c of the pyamqp message?

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 2]
```diff
@@ -5,124 +5,210 @@
...
+        self.connection_string=self.get_from_env("AZURE_SERVICEBUS_CONNECTION_STRING")
+        transport_type=TransportType.AmqpOverWebsocket if arguments.transport_type == 1 else TransportType.Amqp
+        mode=ServiceBusReceiveMode.PEEK_LOCK if arguments.peeklock else ServiceBusReceiveMode.RECEIVE_AND_DELETE
+        self.servicebus_client=ServiceBusClient.from_connection_string(
+            self.connection_string,
+            receive_mode=mode,
+            prefetch_count=self.args.num_messages,
+            max_wait_time=self.args.max_wait_time or None, 
+            transport_type=transport_type,
+            uamqqp_transport=arguments.transport_type,
```
Comment (l0lawrence) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:28]: will these run if we don't have the switch yet b/c of uamqp_transport?

Comment (l0lawrence) [reply sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:28]: should we add it in if not already there so the perftests are happy?

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 3]
```diff
@@ -5,124 +5,210 @@
...
+        async with self.async_servicebus_client.get_queue_sender(self.queue_name) as sender:
+            batch=await sender.create_message_batch()
 
+            for i in range(current_queue_message_count, self.args.preload):
+                try:
+                    batch.add_message(ServiceBusMessage(data))
+                except ValueError:
+                    await sender.send_messages(batch)
+                    print(f"Loaded {i} messages")
+                    batch=await sender.create_message_batch()
```
Comment (l0lawrence) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:84]: nit: spaces batch = await

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 4]
```diff
@@ -5,124 +5,210 @@
...
+
+        async with ServiceBusAdministrationClient.from_connection_string(self.connection_string) as admin_client:
+            topic_properties = await admin_client.get_topic_runtime_properties(self.topic_name)
+            current_topic_message_count = topic_properties.scheduled_message_count
+    
+
+        print(f"The current queue {self.topic_name} has {current_topic_message_count} already")
+
+        async with self.async_servicebus_client.get_topic_sender(self.topic_name) as sender:
+            batch=await sender.create_message_batch()
```
Comment (l0lawrence) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:123]: spacing

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py (hunk 1 above)
Comment (kashifkhan) [reply sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:14]: good catch, I need to get rid of this. That value is in MB, so it was trying to create a queue of size 40 GB so I had to bump it down to a valid value. Since the tests are no longer creating/deleting queues dynamically, dont need it

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py (hunk 2 above)
Comment (kashifkhan) [reply sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:28]: hmm it should noop imo. Its passed in as a kwarg and right now shouldnt be used anywhere

**Review by swathipil — COMMENTED** [review]
lgtm! did we want to perf test other things like peek/etc.?

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py (hunk 2 above)
Comment (swathipil) [reply sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:26]: also there's a typo:
```suggestion
            uamqp_transport=arguments.transport_type,
```

File: sdk/servicebus/azure-servicebus/tests/perf_tests/receive_topic_message_batch.py [hunk 5]
```diff
@@ -5,27 +5,27 @@
 
 import asyncio
 
-from ._test_base import _ReceiveTest
+from ._test_base import _TopicReceiveTest
 
 
-class ReceiveMessageBatchTest(_ReceiveTest):
-    def run_sync(self):
+class ReceiveTopicMessageBatchTest(_TopicReceiveTest):
```
Comment (swathipil) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/receive_topic_message_batch.py:11]: I think all the receive topics should be ReceiveSubscription, since we can only send to topics.

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 6]
```diff
@@ -5,124 +5,208 @@
...
+        self.receiver.close()
+        await self.async_receiver.close()
         await super().close()
+    
+   
 
-
-class _ReceiveTest(_QueueTest):
-    def __init__(self, arguments):
+class _TopicReceiveTest(_ReceiveTest):
```
Comment (swathipil) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:99]: might be good to rename to:
```suggestion
class _SubscriptionReceiveTest(_ReceiveTest):
```

**Comment by kashifkhan** [comment]
> lgtm! did we want to perf test other things like peek/etc.?

I didn't do peek because that is limited to 250 messages max by the service. So it is limited and we wont be able to receive anything more than that per call. 

For the others I left it for as needed. SB doesn't seem to be as throughput driven as EH, so our performance outlook will also be different towards it.  If we feel some capability needs to be there we can add it but largely the ability to send and receive messages in a timely matter is what we should cover while live tests handle the other functionality.

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 7]
```diff
@@ -5,124 +5,208 @@
...
+class _SendTest(BatchPerfTest):
+    def __init__(self, arguments) -> None:
+        super().__init__(arguments)
+        transport_type=TransportType.AmqpOverWebsocket if arguments.transport_type == 1 else TransportType.Amqp
+
+        self.connection_string=self.get_from_env("AZURE_SERVICEBUS_CONNECTION_STRING")
+        self.service_client=ServiceBusClient.from_connection_string(
+            self.connection_string,
+            transport_type=transport_type,
+            uamqp_transport=arguments.transport_type,
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:154]:
```suggestion
            uamqp_transport=arguments.uamqp_transport,
```

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 8]
```diff
@@ -5,124 +5,208 @@
...
+        )
+        self.async_service_client=AsyncServiceBusClient.from_connection_string(
+            self.connection_string,
+            transport_type=transport_type,
+            uamqp_transport=arguments.transport_type,
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:159]:
```suggestion
            uamqp_transport=arguments.uamqp_transport,
```

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 9]
```diff
@@ -5,124 +5,208 @@
...
-        parser.add_argument('--max-wait-time', nargs='?', type=int, help='Max time to wait for messages before closing. Defaults to 0.', default=0)
-        parser.add_argument('--preload', nargs='?', type=int, help='Number of messages to preload. Default is 10000.', default=10000)
+    def add_arguments(parser) -> None:
+        super(_SendTest, _SendTest).add_arguments(parser)
+        parser.add_argument('--message-size', nargs='?', type=int, help='Size of a single message. Defaults to 100 bytes', default=100)
+        parser.add_argument('--batch-size', nargs='?', type=int, help='The number of messages that should be included in each batch. Defaults to 100', default=100)
+        parser.add_argument('--uamqp-transport', action="store_true", help="Switch to use uamqp transport. Default is False (pyamqp).", default=False)
+        parser.add_argument('--transport-type', nargs='?', type=int, help="Use Amqp (0) or Websocket (1) transport type. Default is Amqp.", default=0)
+        parser.add_argument('--no-client-share', action='store_true', help='Create one ServiceClient per test instance.  Default is to share a single ServiceClient.', default=False)
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:173]: Not sure we need this one?

File: sdk/servicebus/azure-servicebus/tests/perf_tests/receive_queue_message_batch.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+import asyncio
+
+from ._test_base import _QueueReceiveTest
+
+
+class ReceiveQueueMessageBatchTest(_QueueReceiveTest):
+    def run_sync(self) -> None:
+        count = 0
+        while count < self.args.num_messages:
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/receive_queue_message_batch.py:14]: I'm wondering if this pattern fits better with the BatchPerfTest, where instead of looping, we could do one `receive_messages` operations and return the number received...

File: sdk/servicebus/azure-servicebus/tests/perf_tests/receive_queue_message_stream.py [hunk 11]
```diff
@@ -23,9 +23,9 @@ def run_sync(self):
                     break
                 count += 1
 
-    async def run_async(self):
+    async def run_async(self) -> None:
         count = 0
-        if self.args.peeklock:
+        if not self.args.peeklock:
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/receive_queue_message_stream.py:28]: This loop is running `complete_message`, so it _is_ peeklock isn't it?

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py (hunk 9 above)
Comment (kashifkhan) [reply sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:173]: No, we can remove this since every run is getting their own client.

File: sdk/servicebus/azure-servicebus/tests/perf_tests/receive_queue_message_stream.py (hunk 11 above)
Comment (kashifkhan) [reply sdk/servicebus/azure-servicebus/tests/perf_tests/receive_queue_message_stream.py:28]: ah I got mixed up on receive & delete vs peeklock and complete. Has to be fixed across all the tests

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 12]
```diff
@@ -5,124 +5,279 @@
...
-
-
-class _ServiceTest(PerfStressTest):
-    service_client = None
-    async_service_client = None
-
-    def __init__(self, arguments):
+class _ReceiveTest():
+    def __init__(self, arguments) -> None:
         super().__init__(arguments)
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:16]: We're calling super, but there's no super?

Comment (annatisch) [reply sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:16]: If this is intended to be a mixin - I would probably remove the constructor and move this code to some kind of setup method.
Having constructors with multiple inheritance can be hard to follow.

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 13]
```diff
@@ -5,124 +5,279 @@
...
+        self.receiver.close()
+        await self.async_receiver.close()
+        await self.close_clients()
+        await super().close()
+    
+    @staticmethod
+    def add_arguments(parser) -> None:
+        super(_QueueReceiveTest, _QueueReceiveTest).add_arguments(parser)
+        parser.add_argument('--message-size', nargs='?', type=int, help='Size of a single message. Defaults to 100 bytes', default=100)
+        parser.add_argument('--num-messages', nargs='?', type=int, help='Number of messages to send or receive. Defaults to 100', default=100)
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:117]:
```suggestion
        parser.add_argument('--num-messages', nargs='?', type=int, help='Number of messages to receive. Defaults to 100', default=100)
```

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 14]
```diff
@@ -5,124 +5,279 @@
...
+        self.receiver.close()
+        await self.async_receiver.close()
+        await self.close_clients()
+        await super().close()
+    
+    @staticmethod
+    def add_arguments(parser) -> None:
+        super(_SubscriptionReceiveTest, _SubscriptionReceiveTest).add_arguments(parser)
+        parser.add_argument('--message-size', nargs='?', type=int, help='Size of a single message. Defaults to 100 bytes', default=100)
+        parser.add_argument('--num-messages', nargs='?', type=int, help='Number of messages to send or receive. Defaults to 100', default=100)
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:149]:
```suggestion
        parser.add_argument('--num-messages', nargs='?', type=int, help='Number of messages to receive. Defaults to 100', default=100)
```

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 15]
```diff
@@ -5,124 +5,279 @@
...
         self.receiver.close()
         await self.async_receiver.close()
+        await self.close_clients()
         await super().close()
+    
+    @staticmethod
+    def add_arguments(parser) -> None:
+        super(_QueueReceiveBatchTest, _QueueReceiveBatchTest).add_arguments(parser)
+        parser.add_argument('--message-size', nargs='?', type=int, help='Size of a single message. Defaults to 100 bytes', default=100)
+        parser.add_argument('--num-messages', nargs='?', type=int, help='Number of messages to send or receive. Defaults to 100', default=100)
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:179]:
```suggestion
        parser.add_argument('--num-messages', nargs='?', type=int, help='Number of messages to receive. Defaults to 100', default=100)
```

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 16]
```diff
@@ -5,124 +5,279 @@
...
+        await self.close_clients()
+        await super().close()
+    
     @staticmethod
-    def add_arguments(parser):
-        super(_ReceiveTest, _ReceiveTest).add_arguments(parser)
+    def add_arguments(parser) -> None:
+        super(_SubscriptionReceiveBatchTest, _SubscriptionReceiveBatchTest).add_arguments(parser)
+        parser.add_argument('--message-size', nargs='?', type=int, help='Size of a single message. Defaults to 100 bytes', default=100)
+        parser.add_argument('--num-messages', nargs='?', type=int, help='Number of messages to send or receive. Defaults to 100', default=100)
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:210]:
```suggestion
        parser.add_argument('--num-messages', nargs='?', type=int, help='Number of messages to receive. Defaults to 100', default=100)
```

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py (hunk 13 above)
Comment (annatisch) [reply sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:117]: We could probably further update this message to better describe how it's used.

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py [hunk 17]
```diff
@@ -5,124 +5,307 @@
...
     
-    async def close(self):
+    async def close(self) -> None:
         self.receiver.close()
         await self.async_receiver.close()
+        await self.close_clients()
         await super().close()
+    
+    @staticmethod
+    def add_arguments(parser) -> None:
```
Comment (annatisch) [inline sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:196]: It looks like this `add_arguments` method could be moved up onto the mixin, as it seems to be generally the same for all the receive tests?

**Review by annatisch — APPROVED** [review]

File: sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py (hunk 17 above)
Comment (kashifkhan) [reply sdk/servicebus/azure-servicebus/tests/perf_tests/_test_base.py:196]: I had it set up like that initially but the call to super was erroring out as its static. Given that none of the other classes have add arguments, I dont need to call it :)
